### ISSUE TITLE
TOS acceptance can prevent saving settings or accessing account dashboard

### ISSUE THREAD

**Issue by csmcneill** [Automattic/woocommerce-payments#5275]
### Describe the bug
We've seen an uptick in interactions recently related to merchants who are unable to:
1. Save their WCPay settings
2. Use the `Manage in Stripe` or `Edit Details` links to edit their account details

We believe that this issue is at least partly related to TOS acceptance.

When assessing the Stripe logs for affected accounts who cannot access their Stripe account, we see the following error:

```
invalid_request_error
Cannot create a login link for an account that has not completed onboarding.
```

### To Reproduce
We've been unable to reproduce this issue 100%.

We can create an account without accepting the TOS by:
1. Starting the KYC process.
2. Exiting before selecting `Agree & Submit` on the last page.

However, in doing this, we are still able to save the account settings and navigate to the Stripe dashboard to adjust the account details (including accepting the TOS).

A variation of this is:
1. Start the KYC process
2. Exit before adding bank account information
3. Navigate to **Payments**.
4. Select the link to finish onboarding
5. Complete enough information to reach the bank account information (e.g. add a SSN for a US account)
6. Exit again before adding bank account information

However, this method still allows us to save account settings and navigate to the Stripe dashboard.

### Actual behavior
When customers experience this issue, they cannot save their settings. Merchants notice that their account is restricted, but they are not redirected to their account when selecting the `Manage in Stripe` or `Edit Details` links in the **Payments** tab. As a result, they don't have many options available to them, and they reach out to support for assistance.

### Potential workaround
We've tried directing users to navigate to https://connect.stripe.com/app/express so that they can log in from there. This, in theory, would allow them to accept the TOS and bypass the errors they are experiencing. We haven't gotten much feedback about whether or not this is a viable workaround.

### Screenshots
![Markup on 2022-12-11 at 10:50:39](https://user-images.githubusercontent.com/82999806/206917124-a6370f79-71df-4161-8940-55d522bd9dd4.png)
![Markup on 2022-12-11 at 10:51:22](https://user-images.githubusercontent.com/82999806/206917126-ef62ddda-4de0-4ed6-aed6-662224e9fdaf.png)

### Expected behavior
If a merchant has not accepted the Stripe TOS, they should still be able to save their WCPay settings. They should also be able to navigate to their Stripe dashboard to accept the terms.

### Additional context
p1670359839324109-slack-C03EE4CPPK5
p1670773517359979-slack-C3NCP7ZJ6
p1670426992319729-slack-C9976E5MJ

**Comment by csmcneill** [comment]
5756445-zen
5770728-zen

**Comment by jessy-p** [comment]
@Automattic/moltres As part of Prioritization, bringing this issue to your radar.

**Comment by daquinons** [comment]
Conversation started here with HEs and Payments OPs p1671024604991259/1670359839.324109-slack-C03EE4CPPK5

**Comment by rashedripon** [comment]
5797927-zen reported `error in saving settings`

**Comment by namiokuzono** [comment]
5799922-zen

**Comment by csmcneill** [comment]
5787704-zen

**Comment by csmcneill** [comment]
5787704-zen

**Comment by rashedripon** [comment]
38827700-hc going to delete the current account and let them onboard again.

**Comment by maxlaf** [comment]
37726295-hc

**Comment by jtobiesen** [comment]
5821920-zen

**Comment by maxlaf** [comment]
5727445-zen

**Comment by csmcneill** [comment]
@Automattic/moltres While we're waiting to hear back from Stripe, could one of y'all please prioritize this issue?

**Comment by johndcoy** [comment]
5836647-zen

**Comment by kaushikasomaiya** [comment]
5837665-zen

> I have to say that I’ve previously started the process of the account creation trough stripe but due to internet connection issue I didn’t complete it.

**Comment by dmallory42** [comment]
> https://github.com/orgs/Automattic/teams/moltres While we're waiting to hear back from Stripe, could one of y'all please prioritize this issue?

Hey @csmcneill, so far we've been unable to reproduce this issue. 

Myself and @oaratovskyi both tried, and we were able to get the account in a state where the onboarding was not completed/TOS was not accepted by leaving the process halfway through. 

But even in this case, the WCPay dashboard still worked, and we were able to accept the TOS by going to `Overview > Payments` and following the prompt to complete account setup (which directs the user to Stripe Express where they can accept TOS, similar to the potential workaround described).

If we're able to reproduce the issues being described where users are seeing an error message when trying to access their dashboard or save their settings, then we'd be able to dig into this further. 

Would it be possible to ask a user who sees this issue for credentials on their site so we can go in and take a look? I think that would be a good next step in solving this issue.

Thanks!

**Comment by oaratovskyi** [comment]
After lots of effort to reproduce it and being unable to p1672152862888179/1670359839.324109-slack-C03EE4CPPK5, I submitted a ticket to Stripe on Dec, 27. In the ticket I asked for tips on how we can reproduce it, I've also provided them with fresh examples. They are still working on it. I'll post here when there are updates from Stripe

**Comment by paulostp** [comment]
Hey folks, I asked for details from the merchant in 5836647-zen and they shared the below. I was able to reproduce on a test site.

> If I tap the “ finish set up" button it’ll take me to that purple page for signing in which I haven’t a clue what to do on that page and then when I go back to check something on my arrow ⬅️ it’ll go away and say I’m not able to access the page or it’ll just say that it’s disabled.

Basically, if after clicking **Finish setup** you don't go ahead and proceed with the Stripe Express account creation - either because you navigated away from stripe.com or your internet connection dropped - the plugin will be installed but unusable. 

You will get dropped here:

![image](https://user-images.githubusercontent.com/69197772/210850098-51773873-cd08-4fe3-8cc0-0e5edf6ea22d.png)

If you navigate to Payments and click "Update" or "Edit details", you will get this:

![image](https://user-images.githubusercontent.com/69197772/210850360-c4b02faa-3998-4be9-95d9-f1c15f1e167a.png)

There's no way to recall that Stripe page and resume the process, so the only solution is reaching out to support and asking for the account to be deleted.

**Comment by oaratovskyi** [comment]
I've just got a response from Stripe that resonates with the information provided by @paulostp above (btw thanks for the feedback!).
> We can confirm that login links only work with accounts which have onboarded. The point where login links work is when an account gets user credentials attached to it. The account in question has not established credentials, which would be the first step of the onboarding process. Once this is established, then login links should work.

I am going to test this scenario using JN site, because I didn't test the case when onboarding is left immediately after clicking Finish Setup.

**Comment by oaratovskyi** [comment]
I was able to reproduce this behavior described by @paulostp. Confirmed that `When customers experience this issue, they cannot save their settings` also happens.

The actual steps to reproduce for me are:
1. Go to Payments and click Finish Setup
2. Accept Jetpack auth
3. See Stripe KYC first screen (with phone number)
4. Go back to payments connect page, it will look like on screen 1 [here](https://github.com/Automattic/woocommerce-payments/issues/5275#issuecomment-1372575369) (Sorry, you are not allowed to access this page)
5. Navigate to Payments and click "Update" or "Edit details" and it will be the same as on screen 2 [here](https://github.com/Automattic/woocommerce-payments/issues/5275#issuecomment-1372575369) (There was a problem redirecting you to the account dashboard. Please try again.)
6. Go to Payments -> Settings and try to save settings. You'll see this
<img width="1565" alt="image" src="https://user-images.githubusercontent.com/79862886/210868902-ba585796-e27d-4b9f-a20c-2cc73abf6be8.png">

**Comment by francoishvz** [comment]
5842249-zd-woothemes

Trying workaround by deleting WCPay account so that merchant can start over.

**Comment by dmallory42** [comment]
Great work, @oaratovskyi  - I was also able to reproduce this using the steps you describe.

My thoughts are that there are two ways we could handle this:

**Treat them as a new user**

Detect when the plugin is in this state, and in that case, treat it as if the user has not yet onboarded (meaning we would show them the same onboarding/"Finish setup" page they'd see if they were a new user). 

This could behave in exactly the same way as if they were a new user, so we'd just create a new account for them when they start the onboarding. I think it would be okay to do this, because at this point, in the user's mind, they haven't created an account yet - they started an onboarding process and left it at a very early stage, they probably didn't even realise an account was created behind the scenes at this point.

**Prompt them to finish onboarding**

Detect when the plugin is in this state, but instead show some other flow to the user - for example, a landing page which says "We noticed you started account setup but didn't complete it yet, please click here to complete the setup of your WCPay account". 

In this case, we'd generate a new account link for the user using the `account_links` API, similar to what we do in progressive onboarding when a user returns to complete their onboarding. This would use the same account they began creating, meaning we don't need to create a new account, but would probably be a little bit more work because we'd need to develop a new flow which would probably require design input.

TLDR: I would probably prefer the first approach, since I think it is simpler to just treat a user who only entered the first stage of the Stripe KYC and then left as if they are a new user and prompt them to just start the KYC process again. But I think it's worth considering both approaches! 

Let me know your thoughts. 🙂

**Comment by jp-imagines** [comment]
Another instance in 5847934-zd-woothemes (with some discussion at p1673193625179709-slack-C3NCP7ZJ6). 

This case had a different error when attempting to generate a login link: `Cannot create an edit link for the account {acct_identifier}, which is not an Express account.`

**Comment by csmcneill** [comment]
39455699-hc

**Comment by oaratovskyi** [comment]
@dmallory42 You've suggested great options to solve this issue 👍  I personally like the second option, mainly because new Stripe accounts won't be created. But I am happy to implement any approach which will work better for our merchants!

**Comment by Mxchael** [comment]
5848782-zen

**Comment by dmallory42** [comment]
> @dmallory42 You've suggested great options to solve this issue 👍 I personally like the second option, mainly because new Stripe accounts won't be created. But I am happy to implement any approach which will work better for our merchants!

I'm happy with either approach 👍 but I'm not sure who the best person would be to make a decision - option A would probably be faster, if we wanted to get this solved quickly - we could even do option A for now and option B as a future improvement?

cc. @daquinons for a second opinion!

**Comment by namiokuzono** [comment]
5852314-zen
5855421-zen

**Comment by jtobiesen** [comment]
5854012-zen

**Comment by lynnjat7** [comment]
39361214-hc

**Comment by francoishvz** [comment]
27842102-hc

Used workaround - deleting WC Pay and starting the process again solved the issue for the merchant.

**Comment by csmcneill** [comment]
Occurred with a BuiltBy partner. Covered in p1673549940816069-slack-C028SDRAAJF

**Comment by retnonindya** [comment]
Another report: 5867470-zd-woothemes

**Comment by daquinons** [comment]
> > @dmallory42 You've suggested great options to solve this issue 👍 I personally like the second option, mainly because new Stripe accounts won't be created. But I am happy to implement any approach which will work better for our merchants!
> 
> I'm happy with either approach 👍 but I'm not sure who the best person would be to make a decision - option A would probably be faster, if we wanted to get this solved quickly - we could even do option A for now and option B as a future improvement?
> 
> cc. @daquinons for a second opinion!

Thanks for the ping, and great work looking at alternatives! 🙏 

Seeing the number of reports I think we need to address it asap with option 1, adding option 2 to our backlog as a possible future improvement if we see that it'd improve the UX.

**Comment by imodouglas** [comment]
16222743-hc

Deleted the account and asked them to set it up again.

**Comment by kaushikasomaiya** [comment]
5865922-zen

**Comment by dsmithweb** [comment]
5869216-zen

**Comment by maxlaf** [comment]
5859870-zen

**Comment by AbhinavSakalle** [comment]
38451567-hc

**Comment by maxlaf** [comment]
5877034-zen

**Comment by francoishvz** [comment]
5879198-zd-woothemes

**Comment by fetenlakhal** [comment]
39857971-hc

**Comment by imodouglas** [comment]
5882876-zen

**Comment by wyter** [comment]
5886580-zen

**Comment by imodouglas** [comment]
39937540-hc

**Comment by dsmithweb** [comment]
5892498-zen

**Comment by jtobiesen** [comment]
5894830-zen

**Comment by theabhig** [comment]
5895393-zen

**Comment by Brianmitchtay** [comment]
5900654-zen

**Comment by paulostp** [comment]
5851170-zen

**Comment by zoemyers** [comment]
[4925881-hc](https://hud.happychat.io/chat/4925881)

**Comment by cintacks** [comment]
Another instance: 5905534-zd-woothemes

**Comment by Brianmitchtay** [comment]
Seeing this in 5908271-zen

**Comment by CGA8C** [comment]
5911645-ZD

**Comment by maxlaf** [comment]
39914847-hc

**Comment by jtobiesen** [comment]
5916889-zen

**Comment by lynnjat7** [comment]
31048515-hc

**Comment by daquinons** [comment]
Thank you for all your reports.

This issue has been fixed in the recently released version 5.4.0 

Please, encourage merchants to upgrade.

**Comment by csmcneill** [comment]
User reported this issue in 6286650-zen. 

They onboarded but didn't accept the TOS, which prevented them from saving their WCPay settings at **Payments > Settings**.

Site is AT, so they are running the latest version of WCPay (5.8.1 at the time of this comment).

Reopening.

**Comment by daquinons** [comment]
Thanks @csmcneill! We are adding the issue back to the backlog and we'll explore it again cc @oaratovskyi

**Comment by Brianmitchtay** [comment]
6591424-zen Merchant is on 6.2.2

**Comment by csmcneill** [comment]
6618560-zen is likely encountering this issue. 

They haven't accepted TOS, which means their account is restricted. When they try to update those details by following the prompt in wp-admin...

![Markup on 2023-08-06 at 19:10:32](https://github.com/Automattic/woocommerce-payments/assets/82999806/8daae229-6050-4fa5-9e2b-95fccac0ec7b)

...they get the following error:

```
There was a problem redirecting you to the account dashboard. Please try again.
```

**Comment by daquinons** [comment]
@csmcneill Thanks again for the ping. We are circling back to this issue. We discovered #6931 which could be the cause for this, and we are giving them bot a high priority.

**Comment by daquinons** [comment]
During the investigation for #6931 we discovered that the issue is isolated in the new onboarding flow.

I've been able to recreate the situation of the restricted account with the old onboarding flow and:
- When abandoning the flow during the first screen, the user is redirected back to the Payments connect screen
- When abandoning the flow after entering the mobile phone but without giving personal info, the user is also redirected to the Payments connect screen
- When submitting personal info and abandoning the flow, the user is redirected to the Payments dashboard account with restricted status.
- The user is prompted to update the business info in Stripe
- I've been able to see that this in fact works and the user is able to accept the ToS and provided any missing data
- After updating it, the account gets back to normal

However, as #6931 shares, we are able to reproduce the error that you are just sharing, but this is only affecting those users that go through the new onboarding flow and leaves it after our hosted flow.

**Comment by daquinons** [comment]
As explained [here](p1691380360401669-slack-CGGCLBN58), we are closing this issue to narrow down the problem in #6931.

- When coming from the original onboarding flow, users can complete the missing steps as you can see in here p1691380360401669-slack-CGGCLBN58
- When coming from the new onboarding flow, users will experience [this issue](https://github.com/Automattic/woocommerce-payments/issues/5275#issuecomment-1667014892) which will be worked on in #6931

**Comment by hobhobuk** [comment]
Issue still exists.
